src/posts/gpu-pipeline/index.md

```diff
@@ -0,0 +1,333 @@
+
+```python
+import kvikio.zarr
+
+with zarr.config.enable_gpu():
+    store = kvikio.zarr.GDSStore(root="/tmp/air-temp.zarr")
+    ds = xr.open_dataset(filename_or_obj=store, engine="zarr")
+    assert isinstance(ds.air.data, cp.ndarray)
+```
+
+This will read the data directly from the Zarr store to GPU memory, significantly reducing I/O latency. This is especially useful for large datasets, as it reduces the amount of data that needs to be transferred between CPU and GPU memory, but requires GPU Direct Storage (GDS) to be enabled on your system.
+
+While decompression still occurs on the CPU, this marks an important step toward enabling fully GPU-native workflows. In the figure below, we show the flowchart of the data loading process with GDS enabled.
+
+![Flowchart-technically decompression is still done on CPUs](/posts/gpu-pipline/flowchart_2.png)
+
+Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:
+
+```python
+import cupy_xarray
+
+ds = xr.open_dataset(filename_or_obj="/tmp/air-temp.zarr", engine="kvikio")
+assert isinstance(ds.air.data, cp.ndarray)
+```
+
+How do these two methods, Zarr (CPU) and kvikio (GPU), compare?
+
+(TODO put in benchmark numbers here).
+
+**Note**: For kvikio performance improvements, you need GPU Direct Storage (GDS) enabled on your system. This is a feature that allows the GPU to access data directly from storage, bypassing the CPU and reducing latency. GDS is supported on NVIDIA GPUs with the [GPUDirect Storage](https://docs.nvidia.com/datacenter/pgp/gds/index.html) feature.
+
+### Step 3: GPU-based decompression with nvCOMP 🚀
+
+For a fully GPU-native workflow, we can let the GPU do all of the work, including decompression. This is where [NVIDIA's nvCOMP](https://developer.nvidia.com/nvcomp) library comes in. nvCOMP provides GPU-accelerated compression and decompression algorithms for various data formats, including Zstandard (Zstd). This means that all steps of data loading including reading, decompressing, and transforming data can be done on the GPU, significantly reducing the time spent on data loading.
+
+![GPU native decompression](/posts/gpu-pipline/flowchart_3.png)
+
+Sending compressed instead of uncompressed data to the GPU means less data transfer overall, reducing I/O latency from storage to device.
+To unlock this, we would need zarr-python to support GPU-based decompression codecs, with one for Zstandard (Zstd) currently being implemented in [this PR](https://github.com/zarr-developers/zarr-python/pull/2863).
+
+Figure above shows benchmark comparing CPU vs GPU-based decompression, with or without GDS enabled using [the data reading benchmark here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/benchmark/era5_zarr_benchmark.py).
+
+![GPU native decompression](/posts/gpu-pipline/zstd_benchmark.png)
```

Comment:
do you know why it's slower with GDS?

Reply:
Potentially GDS compat mode. I ran a new set of tests and updated the figure per our previous meeting.